### PR TITLE
Ensure null-coalescing LHS is evaluated only once

### LINKED ISSUE #12655: The null-coalescing operator evaluates the left-hand operand twice.
# Steps to reproduce

```powershell
function A {
    Write-Host "In A"
    return "something"
}

# Here, A is called twice
$y = (A) ?? "blabla"
```

# Expected behavior

The function `A` should be called only once.
```none
In A
```

# Actual behavior

The function `A` is called twice.
```none
In A
In A
```

# Environment data

```none
Name                           Value
----                           -----
PSVersion                      7.0.0
PSEdition                      Core
GitCommitId                    7.0.0
OS                             Microsoft Windows 10.0.18362
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### PR DESCRIPTION
**Pull request by rjmholt**
# PR Summary

Fixes https://github.com/PowerShell/PowerShell/issues/12655.

The null coalescing operator currently double-evaluates its LHS, which is an issue if it has side effects. This changes evaluates the LHS, saves the result to a temporary variable and then proceeds with the conditional operator using the variable.

This PR also removes a redundant type check for AutomationNull that could never be satisfied due to the implementation of AutomationNull.

**This PR should be considered for servicing in PS 7**

## PR Context

LHS in null coalescing expressions is currently double evaluated. It should only be evaluated once.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

/cc @SeeminglyScience

### CODE REVIEW

**Review by daxian-dbw — APPROVED** [review]

File: src/System.Management.Automation/engine/parser/Compiler.cs [hunk 1]
```diff
@@ -982,19 +982,23 @@ private static Expression Coalesce(Expression left, Expression right)
             {
                 return left;
             }
-            else if (leftType == typeof(AutomationNull))
-            {
-                return right;
-            }
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/parser/Compiler.cs:988]: Why removing this shortcut branch?

Comment (vexx32) [reply src/System.Management.Automation/engine/parser/Compiler.cs:988]: See @SeeminglyScience's comment: https://github.com/PowerShell/PowerShell/issues/12655#issuecomment-628595935

File: test/powershell/Language/Operators/NullConditional.Tests.ps1 [hunk 2]
```diff
@@ -172,6 +188,16 @@ Describe 'NullCoalesceOperations' -Tags 'CI' {
         It 'Lhs is $?' {
             {$???$false} | Should -BeTrue
         }
+
+        It 'Should only evaluate LHS once when it is NOT null' {
+            (Set-TestStateWithStringResult) ?? 'Nothing' | Should -BeExactly 'Test'
+            $testState.Value | Should -Be 1
```
Comment (iSazonov) [inline test/powershell/Language/Operators/NullConditional.Tests.ps1:194]: The test function is used once and creates "side effect".
Please inline the function
```suggestion
            $testState = [pscustomobject]@{ Value = 0 }
            ( & { 'Test'; [void]$testState.Value++} ) ?? 'Nothing' | Should -BeExactly 'Test'
            $testState.Value | Should -Be 1
```

File: test/powershell/Language/Operators/NullConditional.Tests.ps1 [hunk 3]
```diff
@@ -172,6 +188,16 @@ Describe 'NullCoalesceOperations' -Tags 'CI' {
...
+        }
+
+        It 'Should only evaluate LHS once when it IS null' {
+            (Set-TestStateWithNullResult) ?? 'Nothing' | Should -BeExactly 'Nothing'
+            $testState.Value | Should -Be 1
```
Comment (iSazonov) [inline test/powershell/Language/Operators/NullConditional.Tests.ps1:199]: The same.

**Review by iSazonov — APPROVED** [review]

**Pull request merged**